addons/udes_stock_cron/models/stock_picking.py

```diff
@@ -0,0 +1,201 @@
+# -*- coding: utf-8 -*-
+import time
+from psycopg2 import OperationalError, errorcodes
+from collections import defaultdict
+
+from odoo import fields, models, _, api
+from odoo.exceptions import UserError, ValidationError
+
+import logging
+
+_logger = logging.getLogger(__name__)
+
+
+PG_CONCURRENCY_ERRORS_TO_RETRY = (
+    errorcodes.LOCK_NOT_AVAILABLE,
+    errorcodes.SERIALIZATION_FAILURE,
+    errorcodes.DEADLOCK_DETECTED,
+)
+MAX_TRIES_ON_CONCURRENCY_FAILURE = 5
+
+
+class StockPicking(models.Model):
+    _inherit = "stock.picking"
+
+    def _reserve_stock_assign(self):
+        """Perform assign of the pickings at the move level because refactoring
+        may change the pickings.
+        This is a function so it can be extended and/or overriden.
+        """
+        # Assign at the move level disabling refactoring
+        moves = self.mapped("move_lines")
+        moves._action_assign()
+        return moves.mapped("picking_id")
+
+    def reserve_stock(self):
+        """
+        Reserve stock according to the number of reservable pickings.
+        If this method is called on an empty recordset it will attempt to
+        reserve stock for all eligible picking types.  If the recordset is not
+        empty, it will reserve stock for the picks in the recordset. If the
+        recordset is not empty, it is the callers responsibility to make sure
+        that the pickings belong to at most one batch, otherwise this method
+        cannot respect the priority order of pickings, in this case the
+        behaviour of this method is undefined.
```

Comment:
Q. Can we just throw an error in the case  where multiple batches are present in the passed picking record set

Reply:
I think this could be done, or I was thinking you could filter out all the pickings that belong to more than one batch? But as I said in one of the other comments I don't think I should mess with the design too much, and it could be implemented in another story.